### ISSUE TITLE
Unexpected `unused_assignments` at Rust 1.92.0

### ISSUE THREAD

**Issue by dalance** [rust-lang/rust-clippy#16223]
### Summary

After updating to Rust 1.92.0, unexpected `unused_assignments` errors seem to occur.
Probably, variables used through attribute marco are detected as unused.

### Lint Name

unused_assignments

### Reproducer

I tried this code:

```rust
use miette::*;
use thiserror::Error;

#[derive(Error, Debug, Diagnostic)]
#[error("oops!")]
#[diagnostic()]
struct MyBad {
    #[source_code]
    src: NamedSource<String>,
    #[label("This bit here")]
    bad_bit: SourceSpan,
}

fn main() -> Result<()> {
    let src = "source\n  text\n    here".to_string();

    Err(MyBad {
        src: NamedSource::new("bad_file.rs", src),
        bad_bit: (9, 4).into(),
    })?;

    Ok(())
}
```

I saw this happen:

```
warning: value assigned to `src` is never read
 --> src/main.rs:9:5
  |
9 |     src: NamedSource<String>,
  |     ^^^
  |
  = help: maybe it is overwritten before being read?
  = note: `#[warn(unused_assignments)]` (part of `#[warn(unused)]`) on by default

warning: value assigned to `bad_bit` is never read
  --> src/main.rs:11:5
   |
11 |     bad_bit: SourceSpan,
   |     ^^^^^^^
   |
   = help: maybe it is overwritten before being read?

warning: value assigned to `src` is never read
 --> src/main.rs:9:5
  |
9 |     src: NamedSource<String>,
  |     ^^^
  |
  = help: maybe it is overwritten before being read?

warning: `test106` (bin "test106") generated 4 warnings (1 duplicate)
```

I expected to see this happen: no warning

### Version

```text
rustc 1.92.0 (ded5c06cf 2025-12-08)
binary: rustc
commit-hash: ded5c06cf21d2b93bffd5d884aa6e96934ee4234
commit-date: 2025-12-08
host: x86_64-unknown-linux-gnu
release: 1.92.0
LLVM version: 21.1.3
```

**Comment by dalance** [comment]
Sorry, this is not clippy error.

**Comment by chenyukang** [comment]
I think this is related to https://github.com/rust-lang/rust/issues/147648#issuecomment-3647869072
